### ISSUE TITLE
ResolveError: Cannot find module "typeorm/metadata/ColumnMetadata"

### ISSUE THREAD

**Issue by alumni** [oven-sh/bun#7633]
### What version of Bun is running?

1.0.17+5e60861c3

### What platform is your computer?

Darwin 22.6.0 arm64 arm

### What steps can reproduce the bug?

1. `bun init && bun add typeorm@0.3.17`.

2. Either in a script (`index.ts`) or in REPL:
```js
import { ColumnMetadata } from 'typeorm/metadata/ColumnMetadata';
```
or just:
```js
require('typeorm/metadata/ColumnMetadata');
```

### What is the expected behavior?

I would expect be able to import the `ColumnMetadata` class.

### What do you see instead?

REPL:
```
Uncaught ResolveError: Cannot find module "typeorm/metadata/ColumnMetadata" from <repo>/$bun$repl.ts
```

`bun run index.ts`:
```
error: Cannot find module "typeorm/metadata/ColumnMetadata" from "<repo>/index.ts"
```

### Additional information

Seems `bun` expects the extension `.js` to be in the import (even though it's not an ESM module?)

**Comment by Jarred-Sumner** [comment]
TypeORM has disabled implicit file extensions in their package.json. This is, unfortunately, not a bug in Bun and  it is how package.json `"exports"` works - package authors can choose what and how their packages are imported.

Adding the `.js` file extension to the end works:
```js
import { ColumnMetadata } from 'typeorm/metadata/ColumnMetadata.js';
```

as does `require`:
```js
require('typeorm/metadata/ColumnMetadata');
```


Specifically, these lines are relevant: https://github.com/typeorm/typeorm/blob/master/package.json#L35-L36

If you continue to see issues with `require`, let us know (I was unable to reproduce that)

**Comment by alumni** [comment]
Ah probably I simplified too much and only tried the `import` without trying CJS. That's right, I don't get the error with `require` in the minimal example I mentioned above. But it does happen for `require` in the original repo which I tried to switch to bun.

I actually discovered this while trying to run the production build of an existing app (compiled with webpack + tsc running under node). The production build is pure CJS and only has a simple `package.json`. In the source repo for that app, I get:

```
➜  <repo> bun repl
Welcome to Bun v1.0.18
Type ".help" for more information.
[!] Please note that the REPL implementation is still experimental!
    Don't consider it to be representative of the stability or behavior of Bun overall.

> require('typeorm/metadata/ColumnMetadata')
Uncaught ResolveError: Cannot find module "typeorm/metadata/ColumnMetadata" from <repo>/$bun$repl.ts

> require('typeorm/metadata/ColumnMetadata.js')
{ ColumnMetadata: [class ColumnMetadata] }
```

There's nothing special in the repo's `package.json` (nothing related to modules/exports/entrypoints).

There's no `tsconfig.json` (since it's the root of a monorepo), but I have a `tsconfig.base.json` which extends from `@tsconfig/node20/tsconfig.json` with these compiler options:
```json
{
  "allowJs": true,
  "allowSyntheticDefaultImports": true,
  "importHelpers": true,
  "resolveJsonModule": true,
}
```

I was just evaluating bun, there's no intention to use it for existing projects yet (project complexity is mid-high, so a lot of things can break, especially NAPI), but it's pretty impressive so far.
